components/AssetPage/WebGL/GLTFViewer.tsx

```diff
@@ -21,13 +58,104 @@ const GLTFViewer = ({ show, assetID }) => {
     return <div className={styles.wrapper}>No preview available for this model, sorry :(</div>
   }
 
-  const gltfFiles = files['gltf']['4k']
+  // const gltfFiles = files['gltf']['4k']
```

Comment:
switching out to real data doesn't seem to work anymore (probably the hacky try/catch.then())

I will figure it out and push this bit so you can see more realistic way. For local data I just took Camera_01 and unzipped it into `/public/tmpdata/Camera_01_4k.gltf/` which does work.

Reply:
Hi, maybe it's the same issue that I'm having:
The GLTF loader tries to load the textures from the same directory as the gltf file, but if you check `gltfFiles.gltf.include`, the textures urls are actually different.

Example: The GLTF loader tries to load https://dl.polyhaven.com/file/ph-assets/Models/gltf/4k/antique_ceramic_vase_01/textures/antique_ceramic_vase_01_diff_4k.jpg (which does not exist), but the actual url of the file is https://dl.polyhaven.com/file/ph-assets/Models/jpg/4k/antique_ceramic_vase_01/antique_ceramic_vase_01_diff_4k.jpg.

I don't know how you could fix it with useGLTF(), but it looks like the GLTFLoader class has a [.parse()](https://threejs.org/docs/index.html?q=gltfl#examples/en/loaders/GLTFLoader.parse) function where you can specify where textures are located.

Reply:
oh that's really weird.. good point

When I used the local GLTF files served locally in the public folder, all assets loaded fine, but the same GLTF file from the API has CORS issues according to the console errors at least, I didn't check the paths to confirm..

I just switched back to the locally hosted GLTF folder to check and it _was_ in the textures folder, and loads fine locally, but those paths don't match the URLs returned by the API.

I guess the server stores the static files differently to how they're packaged/bundled for download and referenced in the main .gltf file itself.

I'm already experimenting with the other loading methods as useGLTF is a bit weird or I'm using it wrong, maybe it all comes back to the same path issue. I'll look at transforming or doing something to load the real paths tonight..

Thanks for pointing it out, I think it's definitely one of the issues breaking my attempt!

Reply:
> I guess the server stores the static files differently to how they're packaged/bundled for download and referenced in the main .gltf file itself.

That's correct, the downloaded zip bundles all the required files into their expected structure according to the GLTF. The actual storage urls are very different.

It'd be ideal to get it working with the files we have, but if it's a major issue we can just generate packed .glb files that don't have any file references. This would require some work on the backend of course, but it's definitely an option.

Reply:
The keys of the `"include": {}` array returned from our API should exactly match the GLTF's expected file structure:

GLTF file:
```json
"images": [
    {
      "mimeType": "image/jpeg",
      "name": "baseball_01_nor_gl",
      "uri": "textures/baseball_01_nor_gl_4k.jpg" This
    },
    {
      "mimeType": "image/jpeg",
      "name": "baseball_01_diff",
      "uri": "textures/baseball_01_diff_4k.jpg"
    },
    {
      "mimeType": "image/jpeg",
      "name": "baseball_01_rough",
      "uri": "textures/baseball_01_rough_4k.jpg"
    }
  ]
  ```
  
  This is our API data:
  ```json
  {
  "gltf": {
    "include": {
      "textures/baseball_01_diff_4k.jpg": {},
      "baseball_01.bin": {},
      "textures/baseball_01_nor_gl_4k.jpg": {} Matches this,
      "textures/baseball_01_rough_4k.jpg": {}
    },
    "size": 2678
  }
}
```

Reply:
@gregzaal thanks makes sense. I will try figure out with the mentioned `.parse()` function or alternative that can do some manipulation on the input data URLs.

I'm sure its solvable client side, even if it means making some updates to a copy of GLTFLoader. Not ideal but probably less work than changing backend (until they do bug fixes and we are stuck on our hacked up old version...).

Reply:
Not quite there yet but the GLTF paths are replaced, I think correctly as files are loading now, and the 'CORS' error was actually the browsers error of file not found... not super helpful, and thanks @dorianzgraggen for pointing it out

```
useEffect(() => {
  fetch(url)
    .then(res => res.json())
    .then(data => {
      console.log(gltfFiles.gltf.include)

      setMappedGLTF(
        {
          ...data,
          images: {
            ...data.images.map((key, idx) => {
                return {
                  ...data.images[idx],
                  uri: gltfFiles.gltf.include[key.uri].url
                }
            })
          }
        }
      )
  });
}, []);

useEffect(() => {
  console.log("output", mappedGLTF)

  if (!mappedGLTF) return;
  const _GLTFLoader = new GLTFLoader();

  _GLTFLoader.parse(JSON.stringify(mappedGLTF), "", console.log)
}, [mappedGLTF]);
```

result as follows for `gltf.images`
```
{
    "0": {
        "mimeType": "image/jpeg",
        "name": "Camera_01_strap_nor_gl",
        "uri": "https://dl.polyhaven.com/file/ph-assets/Models/jpg/4k/Camera_01/Camera_01_strap_nor_gl_4k.jpg"
    },
    ...
```

What I get next is a bit more confusing.. a runtime error about TypedArray length when processing the accessors. I need to dig a bit deeper. It might just be that my onLoad is just a console.log or something stupid.

Reply:
just throwing it out there incase it's helpful for whoever else is trying to mess with the glTF file..

aside from the above code error using {} instead of [] creating an object of images instead of array (oops, trying to be immutable for 'safety' led me to writing it wrong) which is not to glTF spec, digging into why it doesn't load I come across this error in validation in my URL transformed glTF file:

> The uri of the image. **Relative paths are relative to the .gltf file.** Instead of referencing an external file, the uri can also be a data-uri. The image format must be jpg or png.
> Non-relative URI found: 'https://dl.polyhaven.com/file/ph-assets/Models/jpg/4k/Camera_01/Camera_01_strap_diff_4k.jpg'.glTF Validator(NON_RELATIVE_URI)
>Node Exception: Error: ENOENT: no such file or directory, open '/Users/~~/~~/~~/polyhaven.com/public/tmpdata/https:/dl.polyhaven.com/file/ph-assets/Models/jpg/4k/Camera_01/Camera_01_lens_body_nor_gl_4k.jpg'glTF Validator(IO_ERROR)

This is my code editor not browser, so it's referencing the local file system, but maybe the same happens in the browser, if it's expecting (requiring?) relative paths and treats even https:// as if it's just part of a file path, you're not going to have a good time.

https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#uris

Sorry @gregzaal .. needs a bit more digging to have some simple first version up and running.

Reply:
OK so .. setting `path` properly (documentation is patchy) on `GLTFLoader.parse(data, path, callback)` fixes the above mentioned absolute/relative path issue.

After this I was getting an issue with the .bin file, which again was path related. Mapping to the API's returned .bin file fixed it and .. it works.

So it's really messy right now (WIP from debugging), but I have to work, so I will cleanup and add back in the layer functionality bits I had working before locally etc. and refactor a bit.

If you want to try it out it is pulling from the API so you can click around to different models. Textures also seems to work but the camera is in the middle so you have to zoom out to see properly.

As mentioned some models seem a bit weird. Modular industrial pipes has some issue looks like origins are messed up and all parts are centered on origin rather than the offset they are supposed to have.

![modular pipes 3D view](https://user-images.githubusercontent.com/553506/124687127-4cbeac80-df0f-11eb-86ca-1d0cdfe6e1bc.png)

Reply:
> As mentioned some models seem a bit weird. Modular industrial pipes has some issue looks like origins are messed up and all parts are centered on origin rather than the offset they are supposed to have.

Looking a bit further into this I don't think it's a problem with the viewer but how the glTF files are made?

Opening the blender pipe file, the layout is spaced out, while the .glTF file looks as the viewer screenshot, all parts are centered on the origin. 

![blend file pipes](https://user-images.githubusercontent.com/553506/125184732-283e3980-e25b-11eb-90ef-3183360c9129.png)
![gltf file pipes](https://user-images.githubusercontent.com/553506/125184733-2bd1c080-e25b-11eb-8ef3-575ede01fda2.png)

You can see in the outliner that the blender file parts all live in a group, while the glTF are independent objects.

![blend and gltf](https://user-images.githubusercontent.com/553506/125184790-a4d11800-e25b-11eb-9b6b-c4f29d88fe90.png)

Maybe this is all super obvious, sorry, I spend some time digging through the file and trying things out to fix it until realizing I should check the files and see if they are actually different or just displaying wrong.

Reply:
Yep, the gltf was exported before the objects were layed out nicer. Not a problem with your code or gltf in general :)

Reply:
@gregzaal just pushed a couple of commits

- camera 'should' _reasonably_ center and fit the object (didn't test a whole lot yet)
- can solo each texture or unsolo back to original
- can toggle wireframe on top of model (I think that was the preference over showing the model AS wireframe?)

I'm using useReducer for a bit of control but since you didn't want to be able to switch on and off individual items it's a bit overkill. It is however working with it and needs a bit of effort to remove it. I have used simpler useState for other bits like soloing maps and wireframe, so it's use is minimized. Still the pre-processing of the glTF data is kinda helpful for stuff like centering and looping simply through nested/parented objects/groups, so unless I extract that it is quite useful.

Maybe with a little testing it's good enough for first version? There's no switch _between_ HRDIs or better color settings but they should be possible to add once this thing is stable.

The pipe and rock issue outlined above is not the viewer's fault as far as I can tell so not much else to do about it.